File: implementations/go/eapp/channel.go
package eapp

// ChannelMode is the frozen set of §3: one Channel carries exactly one mode, and
// the mode MUST NOT change during the Channel's lifetime (CH-5).
type ChannelMode string

const (
	// ModeRequest is request/response (§3): the mode a request MUST have
	// exactly one answer to, or none.
	ModeRequest ChannelMode = "request"
	// ModeEvent is fire-and-forget (§3): delivery MAY be zero times or many
	// (EV-2, EV-3), and no response is ever expected (EV-1).
	ModeEvent ChannelMode = "event"
	// ModeStream is the cursor-ordered log (§3, §10): at-least-once only.
	ModeStream ChannelMode = "stream"
	// ModeState is the fourth frozen mode (§3). Its *runtime* semantics belong
	// to v3.2.0; at this layer a state Channel is a Channel like the others —
	// created, connected, sent to and consumed — with at-least-once delivery.
	ModeState ChannelMode = "state"
)

// Valid reports whether the mode is one of §3's four.
func (m ChannelMode) Valid() bool {
	switch m {
	case ModeRequest, ModeEvent, ModeStream, ModeState:
		return true
	default:
		return false
	}
}

// ParseChannelMode converts a wire string into a mode.
//
// An unknown mode is EAPP_MODE_INVALID. It is *not* EAPP_UNSUPPORTED: §10.4's
// EAPP_UNSUPPORTED is about a feature a transport cannot carry, whereas an
// unknown mode names nothing this layer could ever implement — and accepting it
// as some default would silently give the caller a Channel of the wrong kind.
func ParseChannelMode(text string) (ChannelMode, error) {
	mode := ChannelMode(text)
	if !mode.Valid() {
		return "", errModeInvalid(
			"channel mode MUST be one of request, event, stream, state (§3); got %q", text)
	}
	return mode, nil
}

// DeliveryGuarantee is §4's delivery semantics.
type DeliveryGuarantee string

const (
	// DeliveryAtMostOnce means a message is delivered at most once: no
	// retransmission, no acknowledgement (DL-3).
	DeliveryAtMostOnce DeliveryGuarantee = "at-most-once"
	// DeliveryAtLeastOnce means a message is delivered at least once, which
	// requires acknowledgement (DL-4) and idempotent consumption (DL-5).
	DeliveryAtLeastOnce DeliveryGuarantee = "at-least-once"
)

// Valid reports whether the guarantee is one of §4's two.
func (d DeliveryGuarantee) Valid() bool {
	return d == DeliveryAtMostOnce || d == DeliveryAtLeastOnce
}

// ParseDeliveryGuarantee converts a wire string into a guarantee.
//
// Anything outside the two is EAPP_DELIVERY_UNSUPPORTED, which is where DL-2
// ("exactly-once MUST NOT 出现在 Core") lands: a caller that asks for exactly-once
// gets a refusal that names delivery, not a silent downgrade to one of the two
// guarantees it did not ask for.
func ParseDeliveryGuarantee(text string) (DeliveryGuarantee, error) {
	guarantee := DeliveryGuarantee(text)
	if !guarantee.Valid() {
		return "", errDeliveryUnsupported(
			"delivery MUST be at-most-once or at-least-once (§4, DL-1, DL-2); got %q", text)
	}
	return guarantee, nil
}

// DefaultDelivery is §4.4's derivation, used when a caller omits `delivery`
// (CC-4): stream and state are at-least-once, everything else at-most-once.
//
// The asymmetry is not arbitrary. A stream or a state Channel exists to be
// resumed from a cursor (ST-2, §15's I4 ruling), and a position that can be lost
// cannot be resumed from — so those two modes cannot be served by a guarantee
// that permits loss.
func (m ChannelMode) DefaultDelivery() DeliveryGuarantee {
	switch m {
	case ModeStream, ModeState:
		return DeliveryAtLeastOnce
	default:
		return DeliveryAtMostOnce
	}
}

// AllowsDelivery reports which guarantees the mode permits (§4.4's table).
func (m ChannelMode) AllowsDelivery(guarantee DeliveryGuarantee) bool {
	switch m {
	case ModeStream, ModeState:
		// DL-6: stream and state are at-least-once only; asking for
		// at-most-once is EAPP_DELIVERY_UNSUPPORTED (CC-5).
		return guarantee == DeliveryAtLeastOnce
	default:
		return guarantee.Valid()
	}
}

// ChannelState is §2.1's lifecycle state.
type ChannelState string

const (
	// ChannelOpen is the state a freshly created Channel is in (CC-8): it exists
	// and is tied to its Binding, but nothing is being served yet.
	ChannelOpen ChannelState = "OPEN"
	// ChannelActive is the state after connect() (CC-8), and the state a
	// Channel returns to when its Binding recovers (CC-2, §2.2's
	// DRAINING --connect--> ACTIVE).
	ChannelActive ChannelState = "ACTIVE"
	// ChannelDraining is what a DORMANT Binding imposes (§2.4): stop taking new
	// messages, let what is in flight finish. It is deliberately not CLOSED,
	// because closing would discard in-flight work that a resume would have
	// completed.
	ChannelDraining ChannelState = "DRAINING"
	// ChannelClosed is terminal (CH-3).
	ChannelClosed ChannelState = "CLOSED"
)

// Valid reports whether the state is one of §2.1's four.
func (s ChannelState) Valid() bool {
	switch s {
	case ChannelOpen, ChannelActive, ChannelDraining, ChannelClosed:
		return true
	default:
		return false
	}
}

// IsTerminal reports whether the state admits no transition (CH-3).
func (s ChannelState) IsTerminal() bool { return s == ChannelClosed }

// Serves reports whether the state accepts new messages and new consumers.
//
// DRAINING does not: §2.4 explains that the whole point of DORMANT ⇒ DRAINING is
// to stop accepting new work ("停止接收新消息，等待在途完成"). A DRAINING Channel
// still *serves* what it already holds, which is why this is a separate question
// from `IsTerminal`.
func (s ChannelState) Serves() bool { return s == ChannelOpen || s == ChannelActive }

// ChannelRef is the only part of a Channel the Composition Core may see (§2.1).
//
// Its two members are exactly the two facts a Binding needs to state the
// relation: which Channel, and which Binding it belongs to. Nothing about mode or
// delivery appears, because the Core MUST NOT define Channel semantics
// (v3.0 CH-1).
type ChannelRef struct {
	// ID is an implementation-chosen handle. Like Binding.ID it is opaque to
	// callers: never parsed, never ordered by.
	ID string `json:"id"`
	// Binding is the id of the Binding this Channel is derived from (CH-1).
	Binding string `json:"binding"`
}

// Channel is §2.1's Channel: its reference plus mode, delivery and state.
//
// `Mode` and `Delivery` are set once at creation and never written again, which
// is how CH-5 and CH-6 ("MUST NOT 在生命周期内改变") are made structural rather
// than promised: there is no code path that assigns them after
// CreateChannel returns. `State` is *derived* on every read (see
// Interaction.channelState) and never stored, for the same reason the Core
// derives BindingState (B-3): a stored state needs every mutation that could
// affect it to remember to fix it up, and §2.4's table has four such mutations.
type Channel struct {
	// ID is the Channel's handle.
	ID string `json:"id"`
	// Binding is the Channel's Binding (CH-1).
	Binding string `json:"binding"`
	// Mode is §3's interactive mode.
	Mode ChannelMode `json:"mode"`
	// Delivery is §4's delivery guarantee.
	Delivery DeliveryGuarantee `json:"delivery"`
	// State is §2.1's lifecycle state, derived per §2.4.
	State ChannelState `json:"state"`
}

// Ref reduces the Channel to the part the Composition Core is allowed to see
// (§2.1, §11's `channelRef`).
func (c Channel) Ref() ChannelRef {
	return ChannelRef{ID: c.ID, Binding: c.Binding}
}

// CreateChannelRequest is §12's input.
//
// `Mode` has no default and MUST be given (CC-3): §3 says a Channel has exactly
// one mode, and a default would let a caller get a Channel of a kind it never
// asked for. `Delivery` may be omitted, and CC-4 says how it is then derived.
type CreateChannelRequest struct {
	// Binding is the id of an existing, non-CLOSED Binding.
	Binding string `json:"binding"`
	// Mode is the Channel's mode (CC-3).
	Mode ChannelMode `json:"mode"`
	// Delivery is optional; empty means "derive it from the mode" (CC-4).
	Delivery DeliveryGuarantee `json:"delivery,omitempty"`
}

File: packages/interaction/src/channel.ts
import { EappError } from '@eapp/core';

/**
 * Channel — EaPP v3.1.0 §2.
 *
 * A ChannelRef is the only part the Composition Core may see (§2.2). Everything else
 * (mode / delivery / state) belongs to this layer and MUST NOT leak downwards.
 *
 * Note that v3.1 §2.1 already lists `'state'` as a ChannelMode. No later layer extends
 * this union.
 */

export type ChannelMode = 'request' | 'event' | 'stream' | 'state';

export type DeliveryGuarantee = 'at-most-once' | 'at-least-once';

export type ChannelState = 'OPEN' | 'ACTIVE' | 'DRAINING' | 'CLOSED';

export interface ChannelRef {
  id: string;
  binding: string;
}

export interface Channel extends ChannelRef {
  mode: ChannelMode;
  delivery: DeliveryGuarantee;
  state: ChannelState;
}

/**
 * A Channel together with the lifecycle operations of §2.4. The bare `Channel` interface
 * stays a pure data view (that is all the Composition Core may see), while anything that
 * actually owns a channel gets these.
 */
export interface ManagedChannel extends Channel {
  connect(): Promise<void>;
  drain(): Promise<void>;
  close(): Promise<void>;
  /** Throws unless the channel is ACTIVE. See the implementation for the three codes. */
  requireActive(operation: string): void;
}

const VALID_DELIVERIES: readonly DeliveryGuarantee[] = ['at-most-once', 'at-least-once'];

/**
 * §4.4 / CC-4: an omitted delivery is derived from the mode.
 *
 * Deliberately NOT tolerant of unknown modes: a typo must fail loudly rather than
 * silently default to a weaker guarantee. The layer checks the mode against
 * `VALID_MODES` before reaching here, so the fallback branch is only reachable for a
 * mode that is already known.
 */
export function defaultDeliveryFor(mode: ChannelMode): DeliveryGuarantee {
  return mode === 'stream' || mode === 'state' ? 'at-least-once' : 'at-most-once';
}

/**
 * §4.4: `stream` and `state` allow only `at-least-once`; the other modes allow either.
 *
 * The second check alone was not enough. It only asks whether the mode and the delivery
 * are *compatible*, so a third value such as `'exactly-once'` was accepted for `event`
 * and carried on the Channel — while DL-1 says delivery MUST be one of the two, and
 * DL-2 says exactly-once MUST NOT appear in Core at all. TypeScript callers could not
 * pass it; the value arrives from a wire or another language, where the type system is
 * not there to help. The conformance harness found this the first time it probed an
 * out-of-range value. Nothing else had.
 */
export function assertDeliveryAllowed(mode: ChannelMode, delivery: DeliveryGuarantee): void {
  if (!VALID_DELIVERIES.includes(delivery)) {
    throw new EappError(
      'EAPP_DELIVERY_UNSUPPORTED',
      `unknown delivery guarantee '${String(delivery)}'`, // DL-1 / DL-2
    );
  }
  if ((mode === 'stream' || mode === 'state') && delivery !== 'at-least-once') {
    throw new EappError(
      'EAPP_DELIVERY_UNSUPPORTED',
      `mode '${mode}' requires 'at-least-once', got '${delivery}'`,
    );
  }
}

export class ChannelImpl implements ManagedChannel {
  readonly id: string;
  readonly binding: string;
  readonly mode: ChannelMode;
  readonly delivery: DeliveryGuarantee;
  #state: ChannelState = 'OPEN';

  constructor(ref: ChannelRef, mode: ChannelMode, delivery: DeliveryGuarantee) {
    assertDeliveryAllowed(mode, delivery);
    this.id = ref.id;
    this.binding = ref.binding;
    this.mode = mode;
    this.delivery = delivery;
  }

  get state(): ChannelState {
    return this.#state;
  }

  /**
   * §2.5: OPEN --connect--> ACTIVE.
   *
   * DRAINING --connect--> ACTIVE is also accepted, because v3.1 §2.4 requires the
   * re-derivation in both directions: a Binding returning to ACTIVE MUST put its
   * Channels back in service, and a DRAINING channel that could never resume would make
   * that impossible.
   */
  async connect(): Promise<void> {
    if (this.#state === 'CLOSED') {
      throw new EappError('EAPP_CHANNEL_CLOSED', `channel ${this.id} is closed`);
    }
    if (this.#state === 'OPEN' || this.#state === 'DRAINING') this.#state = 'ACTIVE';
  }

  /** §2.5: ACTIVE --drain--> DRAINING (stop accepting new work, finish in-flight). */
  async drain(): Promise<void> {
    if (this.#state === 'CLOSED' || this.#state === 'DRAINING') return;
    this.#state = 'DRAINING';
  }

  /** CH-3 / CH-4: CLOSED is terminal and close() is idempotent. */
  async close(): Promise<void> {
    this.#state = 'CLOSED';
  }

  /**
   * Guard for operations that require an ACTIVE channel.
   *
   * The three failure modes are deliberately distinct, because they call for different
   * responses: CLOSED is terminal, DRAINING means "not now, the composition is paused"
   * (§2.4 — stop accepting new work, let in-flight finish), and OPEN means the caller
   * simply forgot to connect().
   */
  requireActive(operation: string): void {
    if (this.#state === 'CLOSED') {
      throw new EappError('EAPP_CHANNEL_CLOSED', `cannot ${operation} on a closed channel`);
    }
    if (this.#state === 'DRAINING') {
      throw new EappError(
        'EAPP_CHANNEL_DRAINING',
        `cannot ${operation} while the channel is draining (its Binding is DORMANT)`,
      );
    }
    if (this.#state === 'OPEN') {
      throw new EappError(
        'EAPP_CHANNEL_INVALID',
        `cannot ${operation} on a channel that has not been connected`,
      );
    }
  }
}
